assets/src/data/model/entity-factory/create.js

```diff
@@ -0,0 +1,701 @@
+ * @param {string} originalFieldName
+ * @param {string} aliasFieldName
+ * @param {Object} opts
+ */
+export const createAliasGetterAndSetter = (
+	instance,
+	originalFieldName,
+	aliasFieldName,
+	opts = {},
+) => {
+	if ( originalFieldName !== aliasFieldName ) {
+		Object.defineProperty( instance, aliasFieldName, {
+			get() {
+				return instance[ originalFieldName ];
+			},
+			set( receivedValue ) {
+				return instance[ originalFieldName ] = receivedValue;
+			},
+			...opts,
+		} );
+	}
+};
+
+/**
+ * A getter creator for a field alias.
+ *
+ * @param {Object} instance
+ * @param {string} originalFieldName
+ * @param {string} aliasFieldName
+ * @param {Object} opts
+ */
+export const createAliasGetter = (
+	instance,
+	originalFieldName,
+	aliasFieldName,
+	opts = {},
+) => {
+	if ( originalFieldName !== aliasFieldName ) {
+		Object.defineProperty( instance, aliasFieldName, {
+			get() {
+				return instance[ originalFieldName ];
+			},
+			...opts,
+		} );
+	}
+};
+
+/**
+ * Creates a fluent setter on the provided instance for the given field name.
+ *
+ * @param {Object} instance
+ * @param {string} fieldName
+ * @param {Object} opts  Options for Object.defineProperty
+ */
+export const createFluentSetter = ( instance, fieldName, opts = {} ) => {
+	Object.defineProperty( instance, 'set' + upperFirst( fieldName ), {
+		get() {
+			return ( receivedValue ) => {
+				instance[ fieldName ] = receivedValue;
+				return instance;
+			};
+		},
+		...opts,
+	} );
+};
+
+/**
+ * Creates initial getters and setters for entities on the provided entity
+ * instance using the given data.
+ * @param {Object} instance
+ * keys on instance.
+ */
+export const createEntityGettersAndSetters = ( instance ) => {
+	const primaryKeys = [];
+	forEach(
+		instance.originalFieldsAndValues,
+		( fieldValue, fieldName ) => {
+			setValidateTypeForField( instance, fieldName, fieldValue );
+			if ( isEntityField( fieldName, instance.schema ) ) {
+				if ( instance.isNew ) {
+					assertValidValueForPreparedField(
+						fieldName,
+						fieldValue,
+						instance
+					);
+				} else {
+					assertValidFieldAndValueAgainstSchema(
+						instance.modelName,
+						fieldName,
+						fieldValue,
+						instance
+					);
+				}
+				setInitialEntityFieldsAndValues(
+					instance,
+					fieldName,
+					fieldValue
+				);
+			}
+			if ( fieldName === '_calculated_fields' ) {
+				setCalculatedFieldAndValues( instance, fieldValue );
+			}
+			if ( fieldName === 'link' ) {
+				createGetter( instance, 'link', fieldValue );
+			}
+			if ( fieldName === '_links' ) {
+				setResources( instance, fieldValue );
+			}
+			if (
+				! instance.isNew &&
+				isPrimaryKeyField( fieldName, instance.schema )
+			) {
+				primaryKeys.push( fieldName );
+			}
+		}
+	);
+	if ( ! instance.isNew && primaryKeys.length ) {
+		createPrimaryKeyFieldGetters( instance, primaryKeys );
+	}
+
+	populatePrimaryKeys( instance );
+	populateMissingFields( instance );
+};
```

Comment:
re `isPrimaryKeyField` calls:  ya that sounds like a great optimization, I'll implement.

re: `setValidateTypeForField`:  it is ONLY called in `setInitialEntityFieldsAndValues` if the fieldValue is undefined.  So its simply a failsafe for some conditions where a field may be undefined depending on what code path comes into `setInitialEntityFieldsAndValues` (i.e. a new entity or populated from existing entity). So this doesn't need changed.

Reply:
Actually I can't refactor `isPrimaryKeyField` as you've suggested.

- `populateMissingFields` has to have the conditional in the callback for the forEach because its making sure we're only processing fields that are NOT a primary key.  I can't move the `isPrimaryKey` outside of the loop.
- `createEntityGettersAndSetters` is similar in that the first forEach loop in there would still need to call `isPrimaryKey` on each iteration.

Keep in mind that although unlikely, its still possible for there to be multiple primary keys for models so I have to account for that.

Although I could use something like memoize to serve up a cached value of the `isPrimaryKey`, I don't think it'd be any more performant than what that function is already doing (a simple object check).  

So after reviewing how I could refactor this to reduce the hits to `isPrimaryKey` (its that non-expensive boolean check that's actually _preventing_ some unnecessary processing).  Maybe you'd like to give a go at doing a refactor including modifying the existing tests?

Reply:
Also if you look at `populateMissingFields`, its ONLY doing the call to `isPrimaryKey` IF the field doesn't already exist in the schema:
```js
if (
	! instance[ fieldName ] &&
	! isPrimaryKeyField( fieldName, instance.schema )
) {
```

Reply:
you've not understood my suggestion then as I fully realize that the calls to `isPrimaryKeyField()` would still have to happen within the loops in `populateMissingFields()` and `createEntityGettersAndSetters()` which is why I said:

> Making the changes above would reduce that to only 20 (once for each of the forEach field loops in setInitialEntityFieldsAndValues() and populateMissingFields()).

so for `populateMissingFields()`, the issue is not that it calls `isPrimaryKeyField()`, it's that it calls `setInitialEntityFieldsAndValues()` which in turn calls `createRawEntityGettersSetters()` and **BOTH** of those functions call `isPrimaryKeyField()`, which could have easily been passed in as a parameter.

I've pushed a commit for the above change for you to review. Tests pass as is without modification

Reply:
Ya I didn't fully grok how your suggested changes would be any different until you did the commit.  I still think the gain is marginal at best but it definitely does reduce some calls to `isPrimaryKeyField()` so am happy with the change. graçias! 👍